ffi/examples/read-table/arrow.c

```diff
@@ -0,0 +1,265 @@
+      g_object_unref(builder);
+      error = NULL;
+      continue;
+    }
+
+    GArrowArray* ret = garrow_array_builder_finish((GArrowArrayBuilder*)builder, &error);
+    if (report_g_error("Can't build string array for parition column", error)) {
+      printf("Giving up on column %s\n", col);
+      g_object_unref(builder);
+      error = NULL;
+      continue;
+    }
+    g_object_unref(builder);
+
+    GArrowField* field = garrow_field_new(col, (GArrowDataType*)garrow_string_data_type_new());
+    GArrowRecordBatch* old_batch = cur_record_batch;
+    cur_record_batch = garrow_record_batch_add_column(old_batch, pos, field, ret, &error);
+    g_object_unref(old_batch);
+    if (cur_record_batch == NULL) {
+      if (error != NULL) {
+        printf("Could not add column at %u: %s\n", pos, error->message);
+        g_error_free(error);
+      }
+    }
+  }
+  return cur_record_batch;
+}
+
+// append a batch to our context
+static void add_batch_to_context(
+  ArrowContext* context,
+  ArrowFFIData* arrow_data,
+  PartitionList* partition_cols,
+  const CStringMap* partition_values)
+{
+  GArrowSchema* schema = get_schema(&arrow_data->schema);
+  GArrowRecordBatch* record_batch = get_record_batch(&arrow_data->array, schema);
+  if (context->cur_filter != NULL) {
+    GArrowRecordBatch* unfiltered = record_batch;
+    record_batch = garrow_record_batch_filter(unfiltered, context->cur_filter, NULL, NULL);
+    // unref the old batch and filter since we don't need them anymore
+    g_object_unref(unfiltered);
+    g_object_unref(context->cur_filter);
+    context->cur_filter = NULL;
+  }
+  record_batch = add_partition_columns(record_batch, partition_cols, partition_values);
+  if (record_batch == NULL) {
+    printf("Failed to add parition columns, not adding batch\n");
+    return;
+  }
+  context->batches =
+    realloc(context->batches, sizeof(GArrowRecordBatch*) * (context->num_batches + 1));
+  context->batches[context->num_batches] = record_batch;
+  context->num_batches++;
+  print_diag(
+    "  Added batch to arrow context, have %i batches in context now\n", context->num_batches);
+}
+
+// convert to a garrow boolean array. can't use garrow_boolean_array_builder_append_values as that
+// expects a gboolean*, which is actually an int* which is 4 bytes, but our slice is a C99 _Bool*
+// which is 1 byte
+static GArrowBooleanArray* slice_to_arrow_bool_array(const KernelBoolSlice slice)
+{
+  GArrowBooleanArrayBuilder* builder = garrow_boolean_array_builder_new();
+  GError* error = NULL;
+  for (uintptr_t i = 0; i < slice.len; i++) {
+    gboolean val = slice.ptr[i] ? TRUE : FALSE;
+    garrow_boolean_array_builder_append_value(builder, val, &error);
+    if (report_g_error("Can't append to boolean builder", error)) {
+      g_object_unref(builder);
+      break;
+    }
+  }
+  if (error == NULL) {
+    GArrowArray* ret = garrow_array_builder_finish((GArrowArrayBuilder*)builder, &error);
+    g_object_unref(builder);
```

Comment:
Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/203#discussion_r1634925933, I would suggest making the control flow a bit more obvious, and also reducing indentation:
```c
if (error != NULL) {
  return NULL;
}

if (ret == NULL) {
  ...
}
return (GArrowBooleanArray*)ret;
```